### PR TITLE
AMP-26793: Enable FM rules in API for donor and component funding

### PR DESCRIPTION
**Pull request by vchihai**

### CODE REVIEW

File: amp/WEB-INF/src/org/digijava/module/aim/annotations/interchange/Interchangeable.java [hunk 1]
```diff
@@ -50,7 +50,9 @@
     /**
      *Whether the field is always required, required for non-draft saves, or not required. 
      */
-    String required () default ActivityEPConstants.REQUIRED_NONE;
+    ActivityEPConstants.RequiredValidation required() default ActivityEPConstants.RequiredValidation.NONE;
+    
+    String requiredFmPath() default "";
```
Comment (developster) [inline amp/WEB-INF/src/org/digijava/module/aim/annotations/interchange/Interchangeable.java]: Please add an explanation how `requiredFmPath` and `required` work.

File: amp/WEB-INF/src/org/digijava/module/aim/dbentity/AmpActivityFields.java [hunk 2]
```diff
@@ -103,22 +106,25 @@
...
     //protected String govAgreementNumber;
 
-    @Interchangeable(fieldTitle = ActivityFieldsConstants.AMP_ACTIVITY_ID, importable = false)
+    @Interchangeable(fieldTitle = ActivityFieldsConstants.AMP_ACTIVITY_ID)
     @PermissibleProperty(type={Permissible.PermissibleProperty.PROPERTY_TYPE_ID})
     @VersionableFieldSimple(fieldTitle = "Internal ID", blockSingleChange = true)
     protected Long ampActivityId ;
 
-    @Interchangeable(fieldTitle = ActivityFieldsConstants.AMP_ID, required = "_ALWAYS_", importable = false)
+    @Interchangeable(fieldTitle = ActivityFieldsConstants.AMP_ID)
```
Comment (developster) [inline amp/WEB-INF/src/org/digijava/module/aim/dbentity/AmpActivityFields.java]: Is this correct? Field `ampId` is no longer required?

File: amp/WEB-INF/src/org/digijava/module/aim/annotations/interchange/Interchangeable.java (hunk 1 above)
Comment (developster) [reply amp/WEB-INF/src/org/digijava/module/aim/annotations/interchange/Interchangeable.java]: Which combination of `required` and `requiredFmPath` is invalid?

File: amp/WEB-INF/src/org/digijava/module/aim/dbentity/AmpActivityFields.java [hunk 3]
```diff
@@ -146,7 +153,7 @@
     @VersionableFieldSimple(fieldTitle = "Document Space")
     protected String documentSpace;
 
-    @Interchangeable(fieldTitle = ActivityFieldsConstants.IS_DRAFT, required="_ALWAYS_", importable=false)
+    @Interchangeable(fieldTitle = ActivityFieldsConstants.IS_DRAFT)
```
Comment (developster) [inline amp/WEB-INF/src/org/digijava/module/aim/dbentity/AmpActivityFields.java]: Again, no longer required?

File: amp/WEB-INF/src/org/digijava/module/aim/dbentity/AmpFunding.java [hunk 4]
```diff
@@ -11,28 +13,32 @@
...
 import org.digijava.module.categorymanager.dbentity.AmpCategoryValue;
 import org.digijava.module.categorymanager.util.CategoryConstants;
-import static org.digijava.kernel.ampapi.endpoints.activity.ActivityEPConstants.REQUIRED_ALWAYS;
 
 @TranslatableClass(displayName = "Funding")
 public class AmpFunding implements Serializable, Versionable, Cloneable {
     //IATI-check: not ignored!
     private static final long serialVersionUID = 1L;
     
+    private static final String COMM = String.valueOf(Constants.COMMITMENT);
```
Comment (developster) [inline amp/WEB-INF/src/org/digijava/module/aim/dbentity/AmpFunding.java]: Seems like it is unused.

File: amp/src/test/java/org/digijava/kernel/ampapi/endpoints/activity/FieldsEnumeratorTest.java [hunk 5]
```diff
@@ -142,32 +150,36 @@ public void testComposition() {
...
+        @Interchangeable(fieldTitle = "field_required_fm_visible", requiredFmPath = "fm visible")
         private String fieldRequiredFmEntryVisible;
 
-        @Interchangeable(fieldTitle = "field_required_fm_hidden", required = "fm hidden")
+        @Interchangeable(fieldTitle = "field_required_fm_hidden", requiredFmPath = "fm hidden")
         private String fieldRequiredFmEntryHidden;
+    
+        @Interchangeable(fieldTitle = "disaster_response", fmPath = "fm visible",
+                requiredFmPath = "fm visible", required = ALWAYS)
+        private String fieldRequiredParentFmEntryVisible;
```
Comment (developster) [inline amp/src/test/java/org/digijava/kernel/ampapi/endpoints/activity/FieldsEnumeratorTest.java]: Please add test for case when `requiredFmPath = "fm visible", required = DRAFT`.

Comment (developster) [reply amp/src/test/java/org/digijava/kernel/ampapi/endpoints/activity/FieldsEnumeratorTest.java]: And `requiredFmPath = "fm hidden", required = DRAFT` and `requiredFmPath = "fm hidden", required = SUBMIT`.

File: amp/WEB-INF/src/org/digijava/module/aim/dbentity/AmpActivityFields.java (hunk 2 above)
Comment (vchihai) [reply amp/WEB-INF/src/org/digijava/module/aim/dbentity/AmpActivityFields.java]: required is not used when importable is false

File: amp/WEB-INF/src/org/digijava/module/aim/dbentity/AmpActivityFields.java (hunk 3 above)
Comment (vchihai) [reply amp/WEB-INF/src/org/digijava/module/aim/dbentity/AmpActivityFields.java]: required is not used when importable is false

Comment (nmandrescu) [reply amp/WEB-INF/src/org/digijava/module/aim/dbentity/AmpActivityFields.java]: I recall that we have to specify draft when importing from AMP Offline. I found however that there is special case handled through OFFLINE_REQUIRED_FIELDS (that btw also includes approval status).

**Pull request merged**